alf/algorithms/dynamics_learning_algorithm.py

```diff
@@ -120,23 +119,36 @@ def get_state_specs(self):
         """
         pass
 
+    def predict_step(self, time_step: TimeStep, state: DynamicsState):
+        """Predict the current observation using ``time_step.prev_action``
+            and the feature of the previous observation from ``state``.
+        Args:
+            time_step (TimeStep): input data for dynamics learning
+            state (DynamicsState): state for dynamics learning
+        Returns:
+            AlgStep:
+                outputs: empty tuple ()
+                state (DynamicsState):
+                info (DynamicsInfo):
+        """
+        pass
+
     def train_step(self, time_step: TimeStep, state: DynamicsState):
         """
         Args:
             time_step (TimeStep): input data for dynamics learning
-            state (Tensor): state for dynamics learning (previous observation)
+            state (DynamicsState): state for dynamics learning (previous observation)
         Returns:
-            TrainStep:
+            AlgStep:
                 outputs: empty tuple ()
                 state (DynamicsState): state for training
                 info (DynamicsInfo):
         """
         pass
 
     def calc_loss(self, info: DynamicsInfo):
-        loss = nest.map_structure(torch.mean, info.loss)
         return LossInfo(
-            loss=info.loss, scalar_loss=loss.loss, extra=loss.extra)
+            loss=info.loss, scalar_loss=info.loss, extra=loss.extra)
```

Comment:
Seems still not right. LossInfo.loss and LossInfo.scalar_loss will be added for calculating gradient in Algorithm.update_with_gradient. LossInfo.loss and LossInfo.scalar_loss should contain different loss.
And in the case of DynamicsLearningAlgorithm, we should mask out FIRST step because the state is not valid for FIRST step and we don't need to predict the observation of the FIRST step. So we should mask out FIRST here and assign the mean to scalar_loss so that ``update_with_gradient`` won't do incorrect masking.

Reply:
Now using scalar loss for training to avoid additional masking.
FIRST based masking was done in the DynamicsLearningAlgorithm.